fix: Drop unnecessary OrderBy from QueryGraph

MBkkt opened the pull request:
Order by is dropped when it's not needed (before union/intersects/join/aggregation/order by/etc). So instead of old code when we reassign order by, now we just skip logical plan order by if it's not needed.

The one special case exists, if we found limit, we should re enable order by back.

Follow up:
Ideally we want to distinguish between Ordered TopN and just Unordered TopN, now it's kind of impossible

CC: @pashandor789

mbasmanova commented on axiom/optimizer/ToGraph.h:
```diff
@@ -200,10 +200,15 @@ class ToGraph {
-  // 'currentDt_'. If 'node' does not match  allowedInDt, wraps 'node' in
-  // a new DerivedTable.
-  PlanObjectP makeQueryGraph(
+  DerivedTableP makeStream(
```
Would it be possible to document these methods?

MBkkt replied:
Added this and other comments https://github.com/facebookincubator/axiom/pull/542/commits/15ec906b932b1517d0f283b08ea9cff9e63c4275

mbasmanova commented:
I'm not able to capture the gist of the new design. Would it be possible to describe how the new logic works to transform a tree of logical plan nodes into a tree of derived tables?

mbasmanova commented:
> But in some cases we need to keep current dt for our current operator (join, and potentially union/etc) in such cases it's possible to make transformation that instead of using current dt for input nodes will create new one and put this new dt (with inputs nodes) into old one (with current)

I'm not following. Would you clarify some more? 

Are we traversing the logical plan tree bottom up and fill up a DT? When DT is "full", we start a new one? How do we handle nodes with multiple inputs? Do we create separate DTs for each input, then combine these if possible?

MBkkt commented on axiom/optimizer/ToGraph.cpp:
```diff
@@ -1667,45 +1625,52 @@ void ToGraph::makeUnionDistributionAndStats(
-          return set;
-        }
-      }
+  // TODO: use deducing this lambda when C++23 is available.
```
ok

MBkkt replied:
https://github.com/facebookincubator/axiom/pull/542/commits/15ec906b932b1517d0f283b08ea9cff9e63c4275

fixed

mbasmanova commented on axiom/optimizer/tests/HiveQueriesTest.cpp:
```diff
@@ -157,5 +157,117 @@ TEST_F(HiveQueriesTest, orderOfOperations) {
+                         .tableScan("nation")
+                         .orderBy({"n_nationkey"})
+                         .limit(0, 10)
+                         .as("n1")
```
Do we need these aliases? I believe not.

MBkkt replied:
Yes, I need this, otherwise I don't understand how to write this

`"n1.n_regionkey = r1.r_regionkey",`

MBkkt replied:
You're right, I missed that they have different prefix

mbasmanova reviewed (commented):
PR description mentions a number of improvements. Would it be possible to add tests for these?

mbasmanova commented on axiom/optimizer/ToGraph.cpp:
```diff
@@ -1481,34 +1452,25 @@ PlanObjectP ToGraph::addFilter(const lp::FilterNode* filter) {
-}
-
-PlanObjectP ToGraph::addLimit(const lp::LimitNode& limitNode) {
+void ToGraph::addLimit(const lp::LimitNode& limit) {
```
This is a nice refactoring, but makes it hard to understand material changes. Would it be possible to extract it into a separate PR?

mbasmanova commented on axiom/optimizer/ToGraph.h:
```diff
@@ -200,10 +200,34 @@ class ToGraph {
+
+  // Helper function to call makeQueryGraph but finishes currentDt_
+  // if it will have limit.
+  // This is useful for cases where we want to apply limit before our node.
```
I believe limit must be the last op added to a DT. Hence, we always want to start a new DT after limit, no?

MBkkt commented:
@mbasmanova I added test for main changes 

About flatten union I think existing test cover this (they're adjusted, that's why I changed PlanMatcher a little)

About sropped order, It's tested in my tests (`(... limit order by) join ...`

If you have thoughts about some concrete tests, can you share them, I will try to add such tests

MBkkt commented on axiom/optimizer/ToGraph.h:
```diff
@@ -200,10 +200,34 @@ class ToGraph {
+  // but removes order by from currentDt_.
+  // This is useful for cases where we want to ignore the order
+  // Example: set operations, join, write, aggregation inputs.
+  DerivedTableP makeUnordered(
```
It's same as result of makeQueryGraph.
As for makeQueryGraph I wrote both cases null and not null

MBkkt commented on axiom/optimizer/tests/HiveQueriesTest.cpp:
```diff
@@ -157,5 +157,117 @@ TEST_F(HiveQueriesTest, orderOfOperations) {
+            .hashJoin(
+                core::PlanMatcherBuilder().tableScan("region").topN().build())
+            .build();
+    ASSERT_TRUE(matcher->match(plan));
```
ok

MBkkt replied:
https://github.com/facebookincubator/axiom/pull/542/commits/35b20d724979aeb5c8a6c82118acd859457c17b1

MBkkt commented on axiom/optimizer/ToGraph.h:
```diff
@@ -200,10 +200,34 @@ class ToGraph {
+  // Otherwise returns the "outer" DerivedTable, which needs to be
+  // finalized by the caller.
+  // This is useful for cases where we want to control from caller
+  // do we need to drop order by/limit/aggregation or not.
```
There's no such cases, comment adjusted, see https://github.com/facebookincubator/axiom/pull/542/commits/35b20d724979aeb5c8a6c82118acd859457c17b1

MBkkt commented on axiom/optimizer/tests/HiveQueriesTest.cpp:
```diff
@@ -157,5 +157,117 @@ TEST_F(HiveQueriesTest, orderOfOperations) {
+  auto referencePlan = exec::test::PlanBuilder(planNodeIdGenerator)
+                           .tableScan("nation", getSchema("nation"))
+                           .orderBy({"n_nationkey"}, false)
+                           .limit(0, 10, false)
```
https://github.com/facebookincubator/axiom/pull/542/commits/35b20d724979aeb5c8a6c82118acd859457c17b1

mbasmanova commented:
Still trying to wrap my head around the new design. 

If we consider only linear trees (no branches / joins) then it makes sense to fill up a DT while traversing the plan bottom up. Before adding a node, check if it "fits" in current DT and if not, finalize current DT and start a new one.

This doesn't quite work for joins. Let's consider an example:

```
- 4: join
   - 3: agg
     -  2: scan
   - 1: scan
```

We start populating a DT with scan 1, then we add scan 2, then add agg 3, then add join 4. They all end up in the same DT, which is incorrect. The agg 3 must be in a separate DT. Is this the bug that triggered this PR? Just want to double check my understanding of the issue with existing code.

Now, it looks like Join introduces a restriction of what can be added to the current DT. This restriction is different from rules that decide whether an operation fits into a DT or not (bottom up). The join says that only tables and filters are allowed. Anything else must be wrapped in a separate DT. I believe the existing allow-in-dt mechanism is supposed to enforce that. Is this not working / buggy?

To summarize, currently there are 2 mechanism that are supposed to work together to cover all cases. On the way down the tree, join node updates allow-in-dt. If a node is not allowed, then it starts a new DT (wrapInDt logic). On the way up, each node checks whether it fits in a DT and, if not, starts a new DT (finalizeDt logic). The difference is wrapInDt creates an empty DT and fills it up using child nodes. FinalizeDT creates a new DT and places current DT into it as a 'table'.

MBkkt commented:
@mbasmanova I can say it this way:

We have two approaches to produce DT

1. finalize then it's necessary (works for linear operators)
2. create new dt for input (needed for joins)

In first approach we cannot do finalize later.
In second approach with my code we also cannot do this later, but we can do it later with current main code.

So in short I tried to make behavior of two approaches consistent which gives some benefits:
1. correct having
2. remove unnecessary order by
3. maybe some other, I'm not sure, needs to think more about what cases we have.

MBkkt committed:
fix: Fix how order by translated from logical plan to query graph

mbasmanova reviewed (commented):
@MBkkt I'll try to understand this PR again. Is PR description up to date? It doesn't seem so. If there is anything you could clarify to help understanding, that would be much appreciated. Otherwise, I'll need to find bandwidth to run and debug the code to try to figure out how it works.

mbasmanova reviewed (commented):
If there is any way to break up this PR into smaller pieces, that would help a lot. Thanks.

mbasmanova commented on axiom/optimizer/DerivedTable.cpp:
```diff
@@ -595,17 +594,79 @@ void DerivedTable::importJoinsIntoFirstDt(const DerivedTable* firstDt) {
+  tables = dt->tables;
+  tableSet = dt->tableSet;
+
+  // TODO: Is setop possible here at all?
```
What is flattenDt for? What does it do? Which code paths does it participate in? How are you thinking about this logic?

MBkkt replied:
flattenDt called when we optimize DT and this DT have single DT in it's tables.

This is possible in two cases:
1) to graph produced such code
2) optimizer imported all joins in single dt

Right now we don't have enough tests for this code + setop, so I'm not sure will or not it called

mbasmanova commented on axiom/optimizer/ToGraph.cpp:
```diff
@@ -1913,166 +1869,160 @@ void ToGraph::makeUnionDistributionAndStats(
+    uint64_t allowedInDt) {
+  auto* outerDt = makeQueryGraph(node, allowedInDt);
+  if (currentDt_->hasOrderBy() && !currentDt_->hasLimit()) {
+    currentDt_->orderKeys.clear();
```
Is there a way to avoid adding OrderBy in the first place? Removing it works, but it is a bit awkward and we also keep the expressions that may have been added specifically for OrderBy.

MBkkt replied:
I'm not sure how, the first idea is handle allowInDt for order by differently.
But I think it doesn't always work.
My suggestion is to think about this separately

MBkkt replied:
I did this as you suggested, now we never reassign orderKeys and never clear them.
We just skip Sort node if it's not necessary

MBkkt commented on axiom/optimizer/DerivedTable.cpp:
```diff
@@ -595,17 +594,79 @@ void DerivedTable::importJoinsIntoFirstDt(const DerivedTable* firstDt) {
+    return;
+  }
+  if (hasLimit() || dt->hasLimit()) {
+    // TODO: Can we combine limits?
```
This isn't true for dt, because if one of them have aggregation, it won't work this way.
Same about cardinality estimation

MBkkt replied:
I removed these changes from this PR, they're not necessary here.
Lets continue this discussion in a separate PR: https://github.com/facebookincubator/axiom/pull/579

MBkkt commented on axiom/optimizer/DerivedTable.h:
```diff
@@ -240,11 +240,6 @@ struct DerivedTable : public PlanObject {
-  /// Number of fully processed leading elements of 'conjuncts'. Set by
-  /// distributeConjuncts(). Used to avoid reprocessing conjuncts in repeated
-  /// calls to distributeConjuncts().
-  int32_t numCanonicalConjuncts{0};
```
I made PR: https://github.com/facebookincubator/axiom/pull/585

MBkkt commented on axiom/optimizer/ToGraph.cpp:
```diff
@@ -179,7 +179,7 @@ ExprCP ToGraph::tryFoldConstant(
     auto exprSet = evaluator_.compile(typedExpr);
     const auto& first = *exprSet->exprs().front();
-    if (first.specialFormKind() != velox::exec::SpecialFormKind::kConstant) {
+    if (!first.isConstant()) {
```
I made PR: https://github.com/facebookincubator/axiom/pull/584

MBkkt commented on axiom/optimizer/ToGraph.cpp:
```diff
@@ -2010,70 +1976,54 @@ DerivedTableP ToGraph::makeQueryGraph(const lp::LogicalPlanNode& logicalPlan) {
+    }
+    return;
+  }
+
```
This is core changes @mbasmanova

MBkkt commented:
@mbasmanova I rebased this PR

mbasmanova commented on axiom/optimizer/ToGraph.cpp:
```diff
@@ -1199,6 +1202,9 @@ AggregationPlanCP ToGraph::translateAggregation(const lp::AggregateNode& agg) {
 }
 
 void ToGraph::addOrderBy(const lp::SortNode& order) {
+  VELOX_DCHECK(currentDt_->orderKeys.empty());
```
nit: perhaps, !currentDt_->hasOrderBy()

mbasmanova reviewed (commented):
@MBkkt Thank you for the optimization. Would you add tests to cover all the cases that are being optimized?

MBkkt commented on axiom/optimizer/ToGraph.h:
```diff
@@ -292,7 +292,7 @@ class ToGraph {
   // group bys as non-top operators, whenever descendents of 'node'
   // are not freely reorderable with its parents' descendents.
-  void wrapInDt(const logical_plan::LogicalPlanNode& node);
+  void wrapInDt(const logical_plan::LogicalPlanNode& node, bool unordered);
```
I think unordered more clear, dropOrderBy just document the code (what is done in the code) but we can read code instead.
unordered document intention (we want unordered subquery), so I think it's better.

Also unordered common word in C++, unordered map/set/etc, so meaning of this really clear.


Although I can modify comment for this function, to clarify it's additionally

MBkkt commented:
> Would you clarify what is Ordered and Unordered TopN?

@mbasmanova 

`select x from order by x limit 10` => provide 10 smallest `x` elements in sorted order, this commonly done with topn (it's ordered topn)
`select * from (select * from order by x limit 10) => provide 10 smallest `x` elements in any order, this commonly done with topn but it doesn't do final sort for heap/partition result/tournament tree (it's unordered topn)

This is common optimization that exists in other query engines
